staging/src/k8s.io/dynamic-resource-allocation/resourceslice/noderesources.go

```diff
@@ -0,0 +1,392 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourceslice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	resourceapi "k8s.io/api/resource/v1alpha2"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	resourceinformers "k8s.io/client-go/informers/resource/v1alpha2"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	// resyncPeriod for informer
+	// TODO (https://github.com/kubernetes/kubernetes/issues/123688): disable?
+	resyncPeriod = time.Duration(10 * time.Minute)
+)
+
+// Controller synchronizes information about resources of one
+// driver with ResourceSlice objects. It currently supports node-local
+// resources. A DRA driver for node-local resources typically runs this
+// controller as part of its kubelet plugin.
+//
+// Support for network-attached resources will be added later.
+type Controller struct {
+	cancel     func(cause error)
+	driverName string
+	owner      Owner
+	kubeClient kubernetes.Interface
+	wg         sync.WaitGroup
+	queue      workqueue.TypedRateLimitingInterface[string]
+	sliceStore cache.Store
+
+	mutex sync.RWMutex
+
+	// When receiving updates from the driver, the entire pointer replaced,
+	// so it is okay to not do a deep copy of it when reading it. Only reading
+	// the pointer itself must be protected by a read lock.
+	resources *Resources
+}
+
+// Resources is a complete description of all resources synchronized by the controller.
+type Resources struct {
+	// NodeResources are resources that are local to one node.
+	NodeResources []*resourceapi.ResourceModel
+}
+
+type Owner struct {
+	APIVersion string
+	Kind       string
+	Name       string
+	UID        types.UID
+}
+
+// StartController constructs a new controller and starts it.
+// If the owner is a v1.Node, then the NodeName field in the
+// ResourceSlice objects is set and used to identify objects
+// managed by the controller. The UID is not needed in that
+// case, the controller will determine it automatically.
+//
+// If a kubeClient is provided, then it synchronizes ResourceSlices
+// with the resource information provided by plugins. Without it,
+// the controller is inactive. This can happen when kubelet is run stand-alone
```

Comment:
> A DRA driver for node-local resources typically runs this
// controller as part of its kubelet plugin.

Do we need to consider the stand-alone mode for a dra plugin?

Reply:
Without kubelet? No, I don't think so.

Reply:
IMO，the dra driver requires kubelet and kube-apisever. so the plugin doesn't support the kubelet  's stand-alone mode, when start the controller, the input parameter shouldn't nil.

Reply:
In other words, can we remove this check?

```
func StartController(ctx context.Context, kubeClient kubernetes.Interface, driverName string, owner Owner, resources *Resources) *Controller {
	if kubeClient == nil {
		return nil
	}
```

Reply:
That is relevant for "stand-alone kubelet". Such a mode of deploying kubelet is still supported, although I vaguely remember a KEP proposing to deprecate that. In that mode, kubelet has no client and this code here must bail out safely.

So we need the if check.

Some unit tests also pass a nil client.

Reply:
> Such a mode of deploying kubelet is still supported, although I vaguely remember a KEP proposing to deprecate that.

https://github.com/kubernetes/enhancements/pull/4581 is proposing deprecating / removing runonce mode from the kubelet, definitely not removing standalone kubelet